[AERIE-1824] - Rework TimeType API

goetzrrGit opened the pull request:
* **Tickets addressed:** AERIE-1824
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
Currently, the eDSL for command authoring has the v1 structure below. The original way isn't too bad but there was some feedback from Basak to make it easier to write.


**Original way:**

```
<command>.absoluteTiming(Temporal.instance)
<command>.relativeTiming(Temporal.duration)
<command>.epochTiming(Temporal.duration)
<command>
```

**New Way:**

With this PR you have multiple ways to specify Timing Types

```
// Absolute
A`2020-060T03:45:19`.ADD_WATER
A(Temporal.Instant.from("2025-12-24T12:01:59Z")).ADD_WATER
A("2020-060T03:45:19").ADD_WATER

// Relative
R`00:15:00`.EAT_BANANA
R(Temporal.Duration.from({ minutes: 15 }).EAT_BANANA
R(00:15:00)EAT_BANANA

// Epoch
E`00:15:30`.PREPARE_LOAF(1,true)
E(Temporal.Duration.from({ minutes: 15, seconds: 30 }).PREPARE_LOAF(1,true)
E(00:15:30).PREPARE_LOAF(1,true)

C.BAKE_BREAD

```

## Verification
All JEST Test Pass on my end. 

To run the test locally:

* Nuke all docker containers
* In Docker Compose, Change `cam` to `none`
* Navigate to `Command-Expansion-Server`
* Enter `nvm use 16` to use node 16
* Run the test `npm run test`
* All the tests should pass

## Documentation
I will update the documentation once this PR is merged.

## Future work
Add more helper methods to make authoring easier.

goetzrrGit committed:
Added TimingType API for commands.

We want to change the API to be like this:

A`2020-001T00:00:00`.ADD_WATER
R`00:00:20         `.PREHEAT_OVEN(act.tbSugar)
E`00:00:30         `.BAKE_BREAD.desc('this does a thing')
C.COOL_DOWN

goetzrrGit committed:
Export new TimingType methods to Authoring language

goetzrrGit committed:
Updated Test to verify new API.

dyst5422 commented on command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts:
```diff
@@ -31,13 +31,13 @@ export type CommandOptions<
 > = { stem: string; arguments: A; metadata?: M } & (
   | {
-      absoluteTime: Temporal.Instant;
+      absoluteTime: String | Temporal.Instant;
```
String is not something you should generally use see https://dev.to/josiasaurel/the-difference-between-string-and-string-in-typescript-4mf2

dyst5422 commented on command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts:
```diff
@@ -146,11 +153,32 @@ export class Command<
+              type: TimingTypes.ABSOLUTE,
+              tag: Command.instantToDoy(
+                (typeof this.absoluteTime === 'string'
+                  ? Command.doyToInstant(this.absoluteTime as DOY_STRING)
```
We should do this conversion in the time functions themselves (A,R,E)

dyst5422 commented on command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts:
```diff
@@ -321,6 +349,57 @@ export class Sequence {
+  return commandsWithTimeValue(duration, TimingTypes.EPOCH_RELATIVE);
+}
+
+function commandsWithTimeValue(
```
We can better type this as
```typescript
function commandsWithTimeValue<T extends TimingTypes>(
  timeValue: T extends TimingTypes.Absolute ? string | Temporal.Instant : string | Temporal.Duration,
  timeType: T
): GeneratedCommands
```

This allows us to enforce that absolute uses strings or instants and relative uses strings or durations

dyst5422 reviewed (changes requested):
I think we can make it a bit more terse using tagged template literals https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates
so we can write it like
```typescript
A`2020-060T03:45:19`.ADD_WATER
R`00:15:00`.EAT_BANANA
E`00:15:30`.PREPARE_LOAF(1,true)
```

We would then support supplying a value that isn't a string using overloads so we can properly type check the variadic inputs.

EX:

```typescript

function A(...args: [TemplateStringsArray, ...string[]]): GeneratedCommands;
function A(timeVal: Temporal.Instant): GeneratedCommands;
function A(timeDOYString: string): GeneratedCommands;
function A(...args: [TemplateStringsArray, ...string[]] | [Temporal.Instant] | [string]): GeneratedCommands {
  let timeVal = timeValue;
  // Handle tagged template format
  if (Array.isArray(args[0])) {
    timeVal = Command.doyToInstant(String.raw(args));
  }
  // Handle string inputs
  else if (typeof args[0] === 'string') {
    timeVal = Command.doyToInstant(args[0]);
  }

  ...do rest of stuff now that we've normalized the time value to a Temporal.Instant


}
```

This allows us to use it in all of the following ways with strong checking

```typescript
A`2020-060T03:45:19`.ADD_WATER

let doyString = '2020-060T03:45:19';
A(doyString).ADD_WATER

let temporalInstant = doyToInstant('2020-060T03:45:19');
A(temporalInstant).ADD_WATER
```

dyst5422 commented on command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts:
```diff
@@ -89,8 +89,31 @@ declare global {
   type F32 = F<32>;
   type F64 = F<64>;
+
+  // @ts-ignore : 'Commands' found in generated code
```
Love the comments here

dyst5422 commented on command-expansion-server/src/lib/codegen/CommandEDSLPreface.ts:
```diff
@@ -263,7 +286,7 @@ export class Command<
   }
 
-  private static hmsToDuration(hms: HMS_STRING): Temporal.Duration {
+  public static hmsToDuration(hms: HMS_STRING): Temporal.Duration {
```
I think we may want to separate these conversions into their own non-exported functions so that they don't become part of the public API?

dyst5422 replied:
@goetzrrGit ☝️

goetzrrGit replied:
Oh I didn't see this. I will look at this now.

goetzrrGit replied:
I pulled the methods out in the last commit.

goetzrrGit committed:
Move time conversion methods to outside the API

* Move hmsToDuration and doyToInstant because they are used by the TimeType helper methods

This will prevent the users from calling these methods when authoring the expansion logic

dyst5422 reviewed (approved):
LGTM
